akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -45,28 +104,167 @@ class AttributesSpec extends StreamSpec {
+        Source.fromGraph(new AttributesSource(Attributes.name("original-name")))
+          .withAttributes(Attributes.name("replaced")) // this actually replaces now
+          .toMat(Sink.head)(Keep.right).withAttributes(Attributes.name("whole-graph"))
+          .run()
+          .futureValue
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "make the attributes on Flow.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.maybe
+          .viaMat(
+            Flow.fromGraph(new AttributesFlow(Attributes.name("original-name")))
+              .withAttributes(Attributes.name("replaced")) // this actually replaces now
+          )(Keep.right)
+          .withAttributes(Attributes.name("source-flow"))
+          .toMat(Sink.ignore)(Keep.left)
+          .withAttributes(Attributes.name("whole-graph"))
+          .run()
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "make the attributes on Sink.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.maybe.toMat(
+          Sink.fromGraph(new AttributesSink(Attributes.name("original-name")))
+            .withAttributes(Attributes.name("replaced")) // this actually replaces now
+        )(Keep.right)
+          .withAttributes(Attributes.name("whole-graph"))
+          .run()
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "use the initial attributes for dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the most specific dispatcher when specified directly around the graph stage" in {
+      val dispatcher =
+        Source.fromGraph(
+          // directly on stage
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher")
+            .addAttributes(ActorAttributes.dispatcher("my-dispatcher")))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the most specific dispatcher when defined on a surrounding composed graph" in {
+      val dispatcher =
+        Source.fromGraph(
+          // on the composed graph
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher"))
+          .map(identity)
+          // this is now for source -> flow
+          .addAttributes(ActorAttributes.dispatcher("my-dispatcher"))
```

Comment:
does this add another async boundary, so that there is one actor for the ThreadNameSnitchingStage and another actor for map(identity) and a third for Sink.head ?

Reply:
Yes, that's what I expect at least.